### PR TITLE
Remove knowledge of specific field marshalers from non-marshalling subsystems

### PR DESCRIPTION
**Pull request by jkoritzinsky**
As part of a general refactoring goal I have of integrating our field marshalling with our parameter marshalling, I've taken a stab at refactoring how non-marshalling portions of the code use NFTs.

- Create a concept of "native field category" which is designed to describe all SystemV classifications, HFA classifications, blittability of the native field representation
- Refactor the `INITFIELDMARSHALER` macro into `InitFieldMarshaler` free function to initialize field marshalers.
- Remove all usages of the `NStructFieldType` enum and the `NFT_*` enum values from files that aren't `fieldmarshaler.h` and `fieldmarshaler.cpp`.

### CODE REVIEW

**Comment by jkoritzinsky** [comment]
@dotnet-bot test this please

File: src/vm/arm/cgencpu.h [hunk 1]
```diff
@@ -103,9 +103,9 @@ EXTERN_C void setFPReturn(int fpSize, INT64 retVal);
 // is guarded by an assert.
 //=======================================================================
 #ifdef _WIN64
-#define MAXFIELDMARSHALERSIZE               40
+#define MAXFIELDMARSHALERSIZE               48
```
Comment (AaronRobinsonMSFT) [inline src/vm/arm/cgencpu.h]: Is there any possible way to compute these values?

File: src/vm/class.cpp [hunk 2]
```diff
@@ -1346,33 +1345,30 @@ CorElementType EEClassLayoutInfo::GetNativeHFATypeRaw()
...
-            fieldType = ((FieldMarshaler_FixedArray *)pFieldMarshaler)->GetElementTypeHandle().GetMethodTable()->GetNativeHFAType();
-            break;
-
-        case NFT_DATE:
-            fieldType = ELEMENT_TYPE_R8;
-            break;
-
-        default:
-            // Not HFA
+        if ((category & NATIVE_FIELD_CATEGORY_FLOAT) == NATIVE_FIELD_CATEGORY_FLOAT)
```
Comment (AaronRobinsonMSFT) [inline src/vm/class.cpp]: Feel free to remove the right-hand side of the `==`. I realize that contradicts my statement about types, but this is a common enough idiom that I think it is obvious how it works.

File: src/vm/fieldmarshaler.h [hunk 3]
```diff
@@ -343,6 +345,18 @@ class FieldMarshaler
...
+    NativeFieldCategory GetNativeFieldCategory() const
+    {
+        LIMITED_METHOD_CONTRACT;
+        return m_nfc;
+    }
+
+    void SetNativeFieldCategory(NativeFieldCategory nfc)
+    {
+        LIMITED_METHOD_CONTRACT;
+        m_nfc = nfc;
```
Comment (AaronRobinsonMSFT) [inline src/vm/fieldmarshaler.h]: Any invariant we can assert here? e.g. initial state of `m_nfc`?

File: src/vm/fieldmarshaler.h [hunk 4]
```diff
@@ -463,6 +478,7 @@ class FieldMarshaler
     RelativeFixupPointer<PTR_FieldDesc> m_pFD;      // FieldDesc
     UINT32           m_dwExternalOffset;    // offset of field in the fixed portion
     NStructFieldType m_nft;
+    NativeFieldCategory m_nfc;
```
Comment (AaronRobinsonMSFT) [inline src/vm/fieldmarshaler.h]: Ah ha! This is the reason for the increase in `MAXFIELDMARSHALERSIZE`. Please try and use `static_assert()` to validate this.

File: src/vm/nativefieldcategory.h [hunk 5]
```diff
@@ -0,0 +1,29 @@
+enum NativeFieldSubcategory
+{
+    NATIVE_FIELD_CATEGORY_FLOAT = 1 << 10,
+    NATIVE_FIELD_SUBCATEGORY_NESTED = 1 << 11,
+#ifdef FEATURE_COMINTEROP
```
Comment (AaronRobinsonMSFT) [inline src/vm/nativefieldcategory.h]: I don't see any reason to do this for an enum value. You can remove the `#ifdef`.

File: src/vm/arm/cgencpu.h (hunk 1 above)
Comment (jkoritzinsky) [reply src/vm/arm/cgencpu.h]: Maybe with `constexpr`? The main issue is that we'd want to use all of the field marshalers and compute the max size, which would get really long. I'd be totally fine doing that though if we want to.

File: src/vm/fieldmarshaler.h (hunk 3 above)
Comment (jkoritzinsky) [reply src/vm/fieldmarshaler.h]: Since `m_nfc` is set as part of the `INITFIELDMARSHALER` macro (the only place field marshalers are created), I think it's fine the way it is right now. It's matching the m_nft field's initialization pattern.

File: src/vm/fieldmarshaler.h (hunk 4 above)
Comment (AaronRobinsonMSFT) [reply src/vm/fieldmarshaler.h]: Oh. Yay! All good then.

File: src/vm/arm/cgencpu.h (hunk 1 above)
Comment (AaronRobinsonMSFT) [reply src/vm/arm/cgencpu.h]: The fact that static asserts exist makes things okay for me.

File: src/vm/fieldmarshaler.h (hunk 3 above)
Comment (AaronRobinsonMSFT) [reply src/vm/fieldmarshaler.h]: I disagree. This is a public function - that really shouldn't be - that should only be called once. I would assert that its initial value is set so we don't call this twice. Unless for some reason this isn't being initialized during construction.

File: src/vm/amd64/cgencpu.h [hunk 6]
```diff
@@ -110,7 +110,7 @@ EXTERN_C void FastCallFinalizeWorker(Object *obj, PCODE funcPtr);
 // as large as the largest FieldMarshaler subclass. This requirement
 // is guarded by an assert.
 //=======================================================================
-#define MAXFIELDMARSHALERSIZE                   40
+#define MAXFIELDMARSHALERSIZE                   48
```
Comment (jkotas) [inline src/vm/amd64/cgencpu.h]: These structures are allocated eagerly today by the type loader for nearly every struct and some classes too, even if they have nothing to do with interop. Is it worth it to allocate 8 bytes more per field just for this cleanup? It does not sound very pay-for-play.

I know that none of this particularly efficient today, but it would be useful to not make it even less efficient.

Comment (AaronRobinsonMSFT) [reply src/vm/amd64/cgencpu.h]: That is a good point - I didn't realize they were allocated so early. It is also interesting that the size increases by 8 bytes with just an enum increase - considering the enum defaults to 32-bits. I am going to assume this is because of some padding requirement.

@jkoritzinsky This change might get a bit more complicated, but I think it might be useful to do some investigation on the layout of the `FieldMarshaler` classes and try to determine if we can use bits that are currently unused in the current layout.

Comment (jkotas) [reply src/vm/amd64/cgencpu.h]: > the size increases by 8 bytes with just an additional enum

That's because of alignment.

Refactoring to make FieldMarshalers allocated on demand would help address my concern too...

Comment (AaronRobinsonMSFT) [reply src/vm/amd64/cgencpu.h]: >That's because of alignment.

Yeah, there might be other padding we can use in this struct. Alternatively, we could use templates for this and sacrifice some code section of the binary in lue of making local members? This could be done for both `NStructFieldType` and `NativeFieldCategory`. Unsure how much code churn that would be.

Yeah I am not sure how easy that would be to change this to allocate on demand. There are a lot of levels here.

@jkoritzinsky The allocation in question is here: https://github.com/dotnet/coreclr/blob/master/src/vm/methodtablebuilder.cpp#L12148-L12149

Comment (jkoritzinsky) [reply src/vm/amd64/cgencpu.h]: It would be a lot more work to make it load on demand.

One thing we could do is change the `NStructFieldType`, `NativeFieldSubcategory`, and `NativeFieldCategory` enumerations only 2-bytes wide each (which would bring the size back down to what it was before. There are a relatively small number of NFT_* types, so limiting that enum value to 16-bits wouldn't be a problem at all. And if we don't want to make the enums 2-bytes wide each, we could just use the higher bits of the `m_nft` field for the categories instead of adding a new field.

File: src/vm/fieldmarshaler.h (hunk 3 above)
Comment (jkoritzinsky) [reply src/vm/fieldmarshaler.h]: Added an assert.

File: src/vm/amd64/cgencpu.h (hunk 6 above)
Comment (jkoritzinsky) [reply src/vm/amd64/cgencpu.h]: I've made the size of each of the enums only 2-bytes wide. So, the size of the largest `FieldMarshaler` has dropped back down to what it was before.

File: src/vm/nativefieldcategory.h [hunk 7]
```diff
@@ -1,29 +1,23 @@
-enum NativeFieldSubcategory
+enum NativeFieldSubcategory : short
```
Comment (AaronRobinsonMSFT) [inline src/vm/nativefieldcategory.h]: This is a really interesting enum. I am not a fan of this kind of splitting concepts across this kind of boundary since it is one step away from doing this entirely with macros. What logic are you trying to handle here?

Comment (jkoritzinsky) [reply src/vm/nativefieldcategory.h]: The idea with separating out the subcategories is that some concepts are relevant to groups of the native field categories (like the FLOAT category is for R4, R8, and DATE categories and the NESTED group for NESTED_TYPE and IN_PLACE_ARRAY categories).

The groups can be used in different situations to collapse the category checking when possible (i.e. checking if a type is an HFA type requires knowing if a type has a nested type, and checking if a type is an SSE type in the SystemV ABI requires knowing if a type is a FLOAT).

Basically it's supposed to be a set of more abstract flags to group the categories below. I'm thinking I should probably swap the naming.

File: src/vm/nativefieldflags.h [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    // The field is marshalled as an in-place (by-value) array.
+    NATIVE_FIELD_CATEGORY_IN_PLACE_ARRAY = NATIVE_FIELD_SUBCATEGORY_ARRAY | NATIVE_FIELD_SUBCATEGORY_NESTED,
+    // The field should be treated like an integer for the purposes of ABIs.
+    NATIVE_FIELD_CATEGORY_INTEGER_LIKE = NATIVE_FIELD_SUBCATEGORY_INTEGER,
+    // The field is being marshaled as a SAFEARRAY*
+    NATIVE_FIELD_CATEGORY_SAFE_ARRAY = NATIVE_FIELD_SUBCATEGORY_COM_ONLY,
+    // The field is being marshaled to a COM interface pointer.
+    NATIVE_FIELD_CATEGORY_INTERFACE_TYPE = NATIVE_FIELD_SUBCATEGORY_COM_ONLY | 0x1,
+    // The field is a structure that is only valid in COM scenarios.
+    NATIVE_FIELD_CATEGORY_COM_STRUCT = NATIVE_FIELD_SUBCATEGORY_COM_ONLY | NATIVE_FIELD_SUBCATEGORY_INTEGER,
```
Comment (AaronRobinsonMSFT) [inline src/vm/nativefieldflags.h]: `NATIVE_FIELD_SUBCATEGORY_INTEGER` - I don't understand how or why something categorized as "COM struct" should also be tagged with an integer. How should I be thinking about this?

File: src/vm/nativefieldflags.h [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+// See the LICENSE file in the project root for more information.
+
+// Describes specific categories/subcategories of native fields.
+enum NativeFieldFlags : short
+{
+    // The native representation of the field is a floating point field.
+    NATIVE_FIELD_SUBCATEGORY_FLOAT = 1 << 8,
+    // The field has a nested MethodTable* (i.e. a field of a struct, class, or array)
+    NATIVE_FIELD_SUBCATEGORY_NESTED = 1 << 9,
+    // The field has a type that is only relevant in COM and should never be marshalled on non-COM ABIs.
```
Comment (AaronRobinsonMSFT) [inline src/vm/nativefieldflags.h]: Love me some double negatives, but communicate this in a way that matches the enum name - "The field has a type that is only relevant in COM and requires adherence to the COM ABI."

File: src/vm/nativefieldflags.h [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    // The field is a user-defined type (value class or layout class)
+    NATIVE_FIELD_CATEGORY_NESTED_TYPE = NATIVE_FIELD_SUBCATEGORY_NESTED,
+    // The field is a System.DateTime (marshals to a double on Windows)
+    NATIVE_FIELD_CATEGORY_DATE = NATIVE_FIELD_CATEGORY_R8 | NATIVE_FIELD_SUBCATEGORY_COM_TYPE,
+    // The field is marshalled as an in-place (by-value) array.
+    NATIVE_FIELD_CATEGORY_IN_PLACE_ARRAY = NATIVE_FIELD_SUBCATEGORY_ARRAY | NATIVE_FIELD_SUBCATEGORY_NESTED,
+    // The field should be treated like an integer for the purposes of ABIs.
+    NATIVE_FIELD_CATEGORY_INTEGER_LIKE = NATIVE_FIELD_SUBCATEGORY_INTEGER,
+    // The field is being marshaled as a SAFEARRAY*
+    NATIVE_FIELD_CATEGORY_SAFE_ARRAY = NATIVE_FIELD_SUBCATEGORY_COM_ONLY,
```
Comment (AaronRobinsonMSFT) [inline src/vm/nativefieldflags.h]: `SAFEARRAY` should not be split.

File: src/vm/nativefieldflags.h (hunk 8 above)
Comment (jkoritzinsky) [reply src/vm/nativefieldflags.h]: The more I think about it, I don't think having `NATIVE_FIELD_SUBCATEGORY_INTEGER` here is needed or actually helpful. I'll remove it.

File: src/vm/methodtable.cpp [hunk 11]
```diff
@@ -2671,13 +2670,20 @@ bool MethodTable::ClassifyEightBytesWithNativeLayout(SystemVStructRegisterPassin
...
+        NativeFieldFlags nfc = pFieldMarshaler->GetNativeFieldFlags();
+
+        if (nfc & NATIVE_FIELD_SUBCATEGORY_COM_TYPE)
         {
-            fieldClassificationType = SystemVClassificationTypeInteger;
+            return false;
         }
-        else if (cls == NFT_FIXEDARRAY)
+#ifdef FEATURE_COMINTEROP
+        else if (nfc & NATIVE_FIELD_SUBCATEGORY_COM_ONLY)
```
Comment (jkoritzinsky) [inline src/vm/methodtable.cpp]: @AaronRobinsonMSFT the only distinction between `NATIVE_FIELD_SUBCATEGORY_COM_TYPE` and `NATIVE_FIELD_SUBCATEGORY_COM_ONLY` is whether or not we assert here. This assert existed in the previous version for the `NFT_*` types that now fall under `NATIVE_FIELD_SUBCATEGORY_COM_ONLY`. Do we want to keep the assert here or just return false?

File: src/vm/nativefieldflags.h [hunk 12]
```diff
@@ -20,17 +22,21 @@ enum NativeFieldFlags : short
     // The native representation of the field has so special properties.
```
Comment (AaronRobinsonMSFT) [inline src/vm/nativefieldflags.h]: > has so special properties.

Clarify.

File: src/vm/fieldmarshaler.cpp [hunk 13]
```diff
@@ -1127,10 +1076,69 @@ do                                                      \
...
 }
+
 #ifdef _PREFAST_
 #pragma warning(pop)
 #endif
 
+BOOL CheckIfDisqualifiedFromManagedSequential(CorElementType corElemType, LayoutRawFieldInfo * pRawFieldInfo, MetaSig &fsig)
+{
+    // This type may qualify for ManagedSequential. Collect managed size and alignment info.
+    if (CorTypeInfo::IsPrimitiveType(corElemType))
```
Comment (AaronRobinsonMSFT) [inline src/vm/fieldmarshaler.cpp]: Flip this a return early.

File: src/vm/fieldmarshaler.cpp [hunk 14]
```diff
@@ -1127,10 +1076,69 @@ do                                                      \
...
+    {
+        pRawFieldInfo->m_managedSize = TARGET_POINTER_SIZE;
+        pRawFieldInfo->m_managedAlignmentReq = TARGET_POINTER_SIZE;
+    }
+    else if (corElemType == ELEMENT_TYPE_VALUETYPE)
+    {
+        TypeHandle pNestedType = fsig.GetLastTypeHandleThrowing(ClassLoader::LoadTypes,
+            CLASS_LOAD_APPROXPARENTS,
+            TRUE);
+        if (pNestedType.GetMethodTable()->IsManagedSequential())
```
Comment (AaronRobinsonMSFT) [inline src/vm/fieldmarshaler.cpp]: Flip and return early.

File: src/vm/fieldmarshaler.h [hunk 15]
```diff
@@ -69,6 +69,48 @@ class FieldMarshaler_Exception;
...
+{
+    ParseNativeTypeFlag_None = 0x00,
+    ParseNativeTypeFlag_IsAnsi = 0x01,
+
+#ifdef FEATURE_COMINTEROP
+    ParseNativeTypeFlag_IsWinRT = 0x02,
+#endif // FEATURE_COMINTEROP
+};
+
+inline ParseNativeTypeFlags operator|=(ParseNativeTypeFlags& lhs, ParseNativeTypeFlags rhs)
```
Comment (AaronRobinsonMSFT) [inline src/vm/fieldmarshaler.h]: This isn't the correct signature - `ParseNativeTypeFlags& operator|=(ParseNativeTypeFlags& lhs, ParseNativeTypeFlags rhs)`;

File: src/vm/fieldmarshaler.h [hunk 16]
```diff
@@ -69,6 +69,48 @@ class FieldMarshaler_Exception;
...
+{
+    LIMITED_METHOD_CONTRACT;
+    lhs = static_cast<ParseNativeTypeFlags>(lhs | rhs);
+    return lhs;
+}
+
+BOOL IsFieldBlittable(FieldMarshaler* pFM);
+
+#ifdef _PREFAST_
```
Comment (AaronRobinsonMSFT) [inline src/vm/fieldmarshaler.h]: Since you pulled this into a header, all this `#ifdef` nonsense can be removed from the declaration.

**Comment by jkoritzinsky** [comment]
@dotnet-bot test this please

**Comment by jkoritzinsky** [comment]
@dotnet-bot test Ubuntu arm Cross Checked Innerloop Build and Test
@dotnet-bot test Ubuntu arm Cross Checked no_tiered_compilation_innerloop Build and Test
@dotnet-bot test Ubuntu x64 Checked Innerloop Build and Test
@dotnet-bot test  Ubuntu x64 Checked Innerloop Build and Test (Jit - TieredCompilation=0) 

Hopefully this time Jenkins won't bork itself.

**Comment by jkoritzinsky** [comment]
@dotnet-bot test Tizen armel Cross Checked Innerloop Build and Test

File: src/vm/methodtablebuilder.cpp [hunk 17]
```diff
@@ -11849,6 +11849,134 @@ MethodTableBuilder::GatherGenericsInfo(
...
+        PRECONDITION(CheckPointer(pInternalImport));
+        PRECONDITION(CheckPointer(pPackingSize));
+        PRECONDITION(CheckPointer(pNLTType));
+        PRECONDITION(CheckPointer(pfExplicitOffsets));
+    }
+    CONTRACTL_END;
+    
+    HRESULT hr;
+    ULONG clFlags;
+#ifdef _DEBUG
```
Comment (AaronRobinsonMSFT) [inline src/vm/methodtablebuilder.cpp]: You can remove this poison initialization on debug builds. The VC++ compiler already does this, unsure about clang, maybe?

File: src/vm/methodtablebuilder.cpp [hunk 18]
```diff
@@ -11849,6 +11849,134 @@ MethodTableBuilder::GatherGenericsInfo(
...
+#endif
+    
+    if (FAILED(pInternalImport->GetTypeDefProps(cl, &clFlags, NULL)))
+    {
+        pAssembly->ThrowTypeLoadException(pInternalImport, cl, IDS_CLASSLOAD_BADFORMAT);
+    }
+    
+    if (IsTdAutoLayout(clFlags))
+    {
+        // <BUGNUM>workaround for B#104780 - VC fails to set SequentialLayout on some classes
```
Comment (AaronRobinsonMSFT) [inline src/vm/methodtablebuilder.cpp]: This is a bug number in an internal bug database. Should be removed.

File: src/vm/methodtablebuilder.cpp [hunk 19]
```diff
@@ -11849,6 +11849,134 @@ MethodTableBuilder::GatherGenericsInfo(
...
+    
+    HRESULT hr;
+    ULONG clFlags;
+#ifdef _DEBUG
+    clFlags = 0xcccccccc;
+#endif
+    
+    if (FAILED(pInternalImport->GetTypeDefProps(cl, &clFlags, NULL)))
+    {
+        pAssembly->ThrowTypeLoadException(pInternalImport, cl, IDS_CLASSLOAD_BADFORMAT);
```
Comment (AaronRobinsonMSFT) [inline src/vm/methodtablebuilder.cpp]: Bah. Can you please make a comment below this line that function `ThrowTypeLoadException()` doesn't return.

File: src/vm/methodtablebuilder.cpp (hunk 18 above)
Comment (jkoritzinsky) [reply src/vm/methodtablebuilder.cpp]: Sounds good.

I do wonder though if we still need this workaround. Doesn't hurt to keep it though.

File: src/vm/olevariant.cpp [hunk 20]
```diff
@@ -755,7 +755,7 @@ UINT OleVariant::GetElementSizeForVarType(VARTYPE vt, MethodTable *pInterfaceMT)
         8,                      // VT_UI8
         4,                      // VT_INT 
         4,                      // VT_UINT
-        0,                      // VT_VOID
+        1,                      // VT_VOID
```
Comment (AaronRobinsonMSFT) [inline src/vm/olevariant.cpp]: Why was this needed?

Comment (jkoritzinsky) [reply src/vm/olevariant.cpp]: I had an assert somewhere while I was writing the GetNativeVariantForVarType function that verified that the native size of the returned type was equal to the native size returned from GetElementSizeForVarType and I was using System.Void for VT_VOID, which is technically size 1. I can't seem to find where I put that assert, so I can probably revert this.

Comment (AaronRobinsonMSFT) [reply src/vm/olevariant.cpp]: I am fine leaving this if need be, but `sizeof(void)` in C is undefined so if we are relying on .NET's interpretation of that size please comment as such.

Comment (jkoritzinsky) [reply src/vm/olevariant.cpp]: I decided to revert it for now since I didn't see any need to change it.

**Comment by jkoritzinsky** [comment]
Marking this as "NO MERGE" so I don't accidentally merge this before we're in the 5.0 time frame.

File: src/vm/fieldmarshaler.cpp [hunk 21]
```diff
@@ -435,7 +435,9 @@ VOID ParseNativeType(Module*                     pModule,
             {
                 if (fDefault || ntype == NATIVE_TYPE_STRUCT)
                 {
-                    selectedNft = InitFieldMarshaler<FieldMarshaler_Decimal>(pfwalk->m_FieldMarshaler, NATIVE_FIELD_CATEGORY_NESTED_VALUE_CLASS);
+                    // The decimal type can't be blittable since the managed and native alignment requirements differ.
```
Comment (AaronRobinsonMSFT) [inline src/vm/fieldmarshaler.cpp]: But what about when it is the only field in a struct?

Comment (jkoritzinsky) [reply src/vm/fieldmarshaler.cpp]: It changes the struct's alignment, so any struct containing that struct and something else might have layout differences. We'd basically have to track this throughout the type system to verify that we don't accidentally mark something blittable that isn't.
